tunnel-server/src/tunnel-store/index.ts

```diff
@@ -0,0 +1,78 @@
+import { KeyObject } from 'crypto'
+import { Logger } from 'pino'
+import { arrayMap } from './array-map'
+
+export { activeTunnelStoreKey } from './key'
+
+export type ScriptInjection = {
+  pathRegex?: RegExp
+  src: string
+  async?: boolean
+  defer?: boolean
+}
+
+export type ActiveTunnel = {
+  envId: string
+  clientId: string
+  tunnelPath: string
+  target: string
+  hostname: string
+  publicKey: KeyObject
+  publicKeyThumbprint: string
+  access: 'private' | 'public'
+  meta: Record<string, unknown>
+  inject?: ScriptInjection[]
+}
+
+export class KeyAlreadyExistsError extends Error {
+  constructor(readonly key: string) {
+    super(`key already exists: "${key}"`)
+  }
+}
+
+export type TransactionDescriptor = { txId: number }
+
+export type ActiveTunnelStore = {
+  get: (key: string) => Promise<ActiveTunnel | undefined>
+  getByPkThumbprint: (pkThumbprint: string) => Promise<readonly ActiveTunnel[] | undefined>
+  set: (key: string, value: ActiveTunnel) => Promise<TransactionDescriptor>
+  delete: (key: string, tx?: TransactionDescriptor) => Promise<void>
+}
+
+const txIdGenerator = () => {
+  let nextTxId = 0
+  return {
+    next: () => {
+      const result = nextTxId
+      nextTxId += 1
+      return result
+    },
+  }
+}
+
+export const inMemoryActiveTunnelStore = ({ log }: { log: Logger }): ActiveTunnelStore => {
+  const keyToTunnel = new Map<string, ActiveTunnel & { txId: number }>()
+  const pkThumbprintToTunnel = arrayMap<string, ActiveTunnel>()
+  const txIdGen = txIdGenerator()
+  return {
+    get: async key => keyToTunnel.get(key),
+    getByPkThumbprint: async pkThumbprint => pkThumbprintToTunnel.get(pkThumbprint),
+    set: async (key, value) => {
+      if (keyToTunnel.has(key)) {
+        throw new KeyAlreadyExistsError(key)
+      }
+      const txId = txIdGen.next()
+      log.debug('setting tunnel %s: %j', key, txId, value)
+      keyToTunnel.set(key, Object.assign(value, { txId }))
+      pkThumbprintToTunnel.add(value.publicKeyThumbprint, value)
+      return { txId }
+    },
+    delete: async (key, tx) => {
```

Comment:
Maybe it's worth returning value if the delete was not commited

Reply:
in a generic implementation, absolutely. I just don't see a use for a return value for the specific use case we have, so I think adding it (and testing for it) is not worth it now.

Reply:
It's not a must, but if we want to add visibility/log that the row was indeed deleted after being called from the client. (due to the issue)